hudi-spark-datasource/hudi-spark-common/src/test/java/org/apache/hudi/TestHoodieSchemaUtils.java

```diff
@@ -0,0 +1,286 @@
+ * under the License.
+ */
+
+package org.apache.hudi;
+
+import org.apache.hudi.common.config.HoodieCommonConfig;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.exception.HoodieNullSchemaTypeException;
+import org.apache.hudi.exception.MissingSchemaFieldException;
+import org.apache.hudi.exception.SchemaBackwardsCompatibilityException;
+
+import org.apache.avro.Schema;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class TestHoodieSchemaUtils {
+
+  @Test
+  void testSchemaWithNullField() {
+    Schema withNullfield = createRecord("nullRecord", createPrimitiveField("nullField", Schema.Type.NULL));
+    assertThrows(HoodieNullSchemaTypeException.class,
+        () -> deduceWriterSchema(withNullfield, null));
+  }
+
+  @Test
+  void testSimplePromotionWithComplexFields() {
+    Schema start = createRecord("simple", createPrimitiveField("f", Schema.Type.INT));
+    Schema end = createRecord("simple", createPrimitiveField("f", Schema.Type.LONG));
+    assertEquals(end, deduceWriterSchema(end, start));
+
+    start = createRecord("nested", createNestedField("f", Schema.Type.INT));
+    end = createRecord("nested", createNestedField("f", Schema.Type.LONG));
+    assertEquals(end, deduceWriterSchema(end, start));
+
+    start = createRecord("arrayRec", createArrayField("f", Schema.Type.INT));
+    end = createRecord("arrayRec", createArrayField("f", Schema.Type.LONG));
+    assertEquals(end, deduceWriterSchema(end, start));
+
+    start = createRecord("mapRec", createMapField("f", Schema.Type.INT));
+    end = createRecord("mapRec", createMapField("f", Schema.Type.LONG));
+    assertEquals(end, deduceWriterSchema(end, start));
+  }
+
+  @Test
+  void testAllowedTypePromotions() {
+    Schema.Type[] promotionTypes = new Schema.Type[]{Schema.Type.INT, Schema.Type.LONG, Schema.Type.FLOAT, Schema.Type.DOUBLE, Schema.Type.STRING, Schema.Type.BYTES};
+    Map<Schema.Type, Pair<Integer,Integer>> allowedPromotions = new HashMap<>();
+    //allowedPromotions.key can be promoted to any type in the range allowedPromotions.value
+    allowedPromotions.put(Schema.Type.INT, Pair.of(0, 4));
+    allowedPromotions.put(Schema.Type.LONG, Pair.of(1, 4));
+    allowedPromotions.put(Schema.Type.FLOAT, Pair.of(2, 4));
+    allowedPromotions.put(Schema.Type.DOUBLE, Pair.of(3, 4));
+    allowedPromotions.put(Schema.Type.STRING, Pair.of(4, 4));
+    allowedPromotions.put(Schema.Type.BYTES, Pair.of(5, 5));
```

Comment:
Based on our docs (https://hudi.apache.org/docs/schema_evolution), type promotion can happen between `BYTES` and `STRING`.  Is that not supported any more?

Reply:
that's not what the docs say. We can reverse promote from string -> byte and byte -> string

Reply:
The type promotion table in docs is meant for actual type changes of the values without changing the table schema for compatibility, but not type promotion of fields in the table schema, which need docs update.